ewms_pilot/utils.py

```diff
@@ -0,0 +1,75 @@
+"""Common utilities."""
+
+
+from typing import Any, Callable, Coroutine, TypeVar
+
+import htchirp  # type: ignore[import]
+from typing_extensions import ParamSpec
+
+from .config import LOGGER
+
+T = TypeVar("T")
+P = ParamSpec("P")
+
+
+def chirp(message: str) -> None:
+    """Invoke HTChirp, AKA send a status message to Condor."""
+    try:
+        chirper = htchirp.HTChirp()
+    except ValueError:  # ".chirp.config must be present or you must provide a host and port
+        return
+
+    with chirper as c:
+        LOGGER.info(f"chirping as '{c.whoami()}'")
+        c.set_job_attr("EWMSPilotProcessing", "True")
+        if message:
+            c.ulog(message)
+
+
+def _initial_chirp() -> None:
+    """Send a Condor Chirp signalling that processing has started."""
+    chirp("")
+
+
+def _final_chirp(error: bool = False) -> None:
+    """Send a Condor Chirp signalling that processing has started."""
+    try:
+        chirper = htchirp.HTChirp()
+    except ValueError:  # ".chirp.config must be present or you must provide a host and port"
+        return
+
+    with chirper as c:
+        LOGGER.info(f"chirping as '{c.whoami()}'")
+        c.set_job_attr("EWMSPilotSucess", str(not error))
+
+
+def error_chirp(exception: Exception) -> None:
+    """Send a Condor Chirp signalling that processing ran into an error."""
+    try:
+        chirper = htchirp.HTChirp()
+    except ValueError:  # ".chirp.config must be present or you must provide a host and port
+        return
+
+    with chirper as c:
+        LOGGER.info(f"chirping as '{c.whoami()}'")
+        c.set_job_attr("EWMSPilotError", "True")
+        c.ulog(f"{type(exception).__name__}: {exception}")
+
+
+def async_htchirping(
```

Comment:
Would it be better to make this an async context manager?  Then you wouldn't have to pass in the callable.

Reply:
I prefer the decorator pattern since it requires fewer code changes, and since chirp is optional

Reply:
Ah.  It wasn't clear that this was a decorator.

Reply:
Note that there is https://docs.python.org/3/library/functools.html#functools.wraps
to keep the function docstrings correct, if we ever used those to make docs.